daemon/snap.go

```diff
@@ -338,12 +342,68 @@ func fillComponentInfo(about aboutSnap) []client.Component {
 	return comps
 }
 
-// snapIcon tries to find the icon inside the snap
+// snapIcon tries to find the icon inside the snap.
 func snapIcon(info snap.PlaceInfo) string {
 	found, _ := filepath.Glob(filepath.Join(info.MountDir(), "meta", "gui", "icon.*"))
-	if len(found) == 0 {
+	if len(found) > 0 {
+		return found[0]
+	}
+
+	// Try to find the icon via the Icon= field of a .desktop file for the snap.
+
+	// Find all .desktop files in ${SNAP}/meta/gui
+	desktopFiles, _ := filepath.Glob(filepath.Join(info.MountDir(), "meta", "gui", "*.desktop"))
+	if len(desktopFiles) == 0 {
 		return ""
 	}
 
-	return found[0]
+	// Try <snap>.desktop first
+	if snapDesktop := info.SnapName() + ".desktop"; strutil.ListContains(desktopFiles, snapDesktop) {
+		if iconPath, err := iconPathFromDesktopFile(info, snapDesktop); err == nil {
+			return iconPath
+		}
+	}
+
+	// Try <snap>_<snap>.desktop next
+	if snapAppDesktop := fmt.Sprintf("%s_%s.desktop", info.SnapName(), info.SnapName()); strutil.ListContains(desktopFiles, snapAppDesktop) {
+		if iconPath, err := iconPathFromDesktopFile(info, snapAppDesktop); err == nil {
+			return iconPath
+		}
+	}
+
+	// Try any .desktop file we could find
+	for _, desktopFile := range desktopFiles {
+		if iconPath, err := iconPathFromDesktopFile(info, desktopFile); err == nil {
+			return iconPath
+		}
+	}
+
+	return ""
+}
+
+var iconMatcher = regexp.MustCompile(fmt.Sprintf("^%s(.*)$", regexp.QuoteMeta("Icon=${SNAP}")))
+
+// iconPathFromDesktopFile opens the given filename and searches for a line
+// starting with `Icon=${SNAP}`, returning the remaining icon path prefixed by
+// the snap's mount directory.
+func iconPathFromDesktopFile(info snap.PlaceInfo, filename string) (string, error) {
+	desktopFile, err := os.Open(filename)
+	if err != nil {
+		return "", err
+	}
+
+	scanner := bufio.NewScanner(desktopFile)
+	for scanner.Scan() {
+		line := scanner.Bytes()
+		matches := iconMatcher.FindSubmatch(line)
+		if matches == nil {
+			continue
+		}
+
+		iconPath := filepath.Join(info.MountDir(), strings.TrimSpace(string(matches[1])))
+
+		return iconPath, nil
```

Comment:
This is the icon path from the point of view of the host, which is safe at least from the point of view of mount table.

Given that mount-control exists though, I wonder if we should be more paranoid here. Would I be able to install a snap that has a desktop file that has an `Icon=${SNAP}/Evil.png` with a symlink somewhere outside of snap mount dir?

What if my Icon= points to ${SNAP}/not-evil/icon.png but then $SNAP/not-evil is used to attach content from antoher snap which shares SNAP_COMMON that has icon.png -> /dev/mem?

CC @alexmurray 

I wonder if we are susceptible to content mount attacks? AFAIR firefox was allowed to mount in the outer mount ns and if we are not careful, we might accidentally get into a situation where a snap mounts over $SNAP and can cause an issue in snapd.